### DISCUSSION TITLE
Avoiding dropped values from AsyncPublisher

### DISCUSSION THREAD

**Discussion by jayhickey** [pointfreeco/swift-composable-architecture#2208] [Q&A] [answered]
Say I have a long living effect like this in a reducer:

```swift

EffectTask.run { send in
  for await value in myPublisher.values { // myPublisher.values is of type AsyncPublisher
    await send(.update(value))
  }
}
```

If `myPublisher` emits a new values quickly and before the `await send…` can complete, we will drop updates. This is because AsyncPublishers don't buffer values, dropping updates instead if the code inside the loop takes too long to run (https://developer.apple.com/forums/thread/694726)

I can work around this by adding some buffering to the stream with something like `myPublisher.buffer(size: 1, prefetch: .byRequest, whenFull: .dropOldest)`, but that's very error prone.

Is there a better way to handle this? Ideally I could queue up actions without needing to `await` completion (i.e., TCA `send` semantics before swift concurrency) or add a cancellation handler to the queued action to drop it if a new one is added. Any suggestions?

Thanks!

**Comment by stephencelis** [comment, marked as answer]
@jayhickey If you're using a Combine publisher, you could use `Effect.publisher`:

```swift
return .publisher { 
  myPublisher
   .map { .update($0) }
}
```

This API is not deprecated and we plan to continue to support it in 1.0.

**Reply by jayhickey** [reply]
Thanks! That's exactly what I was looking for. 

Why isn't that in the documentation? I don't see it listed here:
https://pointfreeco.github.io/swift-composable-architecture/0.49.0/documentation/composablearchitecture/effecttask

**Reply by acosmicflamingo** [reply]
Oh wow I totally forgot about that API! Wait, is that API the preferred way? Instead of using the following, where `deviceOrientation` is a Combine publisher?
```swift
for await orientation in deviceOrientation.values {
  
}
```

**Reply by jayhickey** [reply]
@stephencelis slight correction here, `.publisher { }` is not accessible. I keep getting an error that it's internal only. This works, though:

```swift
return EffectPublisher<Action, Never>( 
  myPublisher
   .map { .update($0) }
)
```

It's a bit concerning though that all the documentation says things like

> 'init (_:' will be deprecated in macOS 9999.0: Iterate over 'Publisher.values' in an
'EffectTask.run', instead.

Given this issue with iterating over the publisher values potentially dropping outputs, I don't think that should be the recommendation?

**Reply by mbrandonw** [reply]
@jayhickey Are you running an old version of TCA? You linked to docs from version 0.49.0 and TCA is currently at 0.54.0. 

In any case, `Effect.publisher` is a public and non-deprecated way to accomplish what you want:

https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/effectpublisher/publisher(_:)

@acosmicflamingo One style isn’t preferred over the other. If you want to use publisher APIs you are free to, and if you want to use structured concurrency you can do that too.

**Reply by jayhickey** [reply]
@mbrandonw Oh yeah, looks like my TCA docs bookmark is out of date 😆 And I was on 0.51.0 for my app, time to update to 0.54.0. Thanks for noticing!
